EASY-1065 deduce default file rights from EMD access rights

jo-pol opened the pull request:
<strike>Not yet tested but</strike> @rvanheest seems to need something like the FileAccessCategory class too

jo-pol committed:
EASY-1065 refactored StageFileItem API for visibility / accessibility

jo-pol committed:
EASY-1065 deduce default file rights from EMD access rights

rvanheest commented:
Yup, gonna need that next week when I implement the issues from Heiko's email in the Licence Creator. Would be nice to either put these File Access Rights in some common library or use the 'old' Java objects instead until we port those libraries to Scala...

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/stage/fileitem/FileAccessCategory.scala:
```diff
@@ -0,0 +1,46 @@
+    REQUEST_PERMISSION -> RESTRICTED_REQUEST,
+    NO_ACCESS -> NONE,
+    // not defined in https://github.com/DANS-KNAW/easy-app/blob/master/lib/easy-business/src/main/java/nl/knaw/dans/easy/domain/model/VisibleTo.java
+    ACCESS_ELSEWHERE -> NONE,
```
Are `ACCESS_ELSEWHERE` and `FREELY_AVAILABLE` used anywhere?

jo-pol replied:
Indeed in DatasetImpl.java I only find

```
categories.add (AccessCategory.ANONYMOUS_ACCESS); // 1 published dataset
categories.add (AccessCategory.OPEN_ACCESS_FOR_REGISTERED_USERS); // 2 known users
categories.add (AccessCategory.GROUP_ACCESS); // 4 member of group
categories.add (AccessCategory.REQUEST_PERMISSION);// 8 granted permission
```

But in the unit test I compare the keysets of both maps against the AccessCategory to make sure we don't forget to update these maps whenever a new access category at dataset level is introduced. It is beyond this issue to clean up that legacy code, perhaps I can  map them to `???` so something gets triggered when these values somehow sneaked into some EMD.

lindareijnhoudt replied:
IIRC freely_available was introduced and used for thumbnails only.

jo-pol replied:
Of course OPEN_ACCESS is recently changed into OPEN_ACCESS_FOR_REGISTERED_USERS. I see NO_ACCESS is used as default in `Excel2EasyMetadataXMLTask.java` when it cannot map the value found in some spread-sheet cell. The new split-multi-deposit currently just copies the value into DDM and I'm afraid Ddm2Emd also simply copies the found value.

Concerns that emerge:
- We should create a test in the Sword-v1 module, I hope it somehow rejects either the DDM or the EMD as invalid. 
- What happens further down the line with the new ingest work flow?

jo-pol replied:
What thumbnails are you talking about?

jo-pol replied:
Both stage-dataset as SWORD-v1 use a cross walker instance with a built-in DDM validator. So we won't encounter an invalid dataset-access-category unless someone tampers with the data between staging and ingest.

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/stage/fileitem/FileAccessCategory.scala:
```diff
@@ -0,0 +1,46 @@
+object FileAccessCategory extends Enumeration {
+  type FileAccessCategory = Value
+  val ANONYMOUS, KNOWN, RESTRICTED_REQUEST, RESTRICTED_GROUP, NONE = Value
+
```
It seems `accessibleTo` and `visibleTo` contain the exact same map. Maybe just create the map once?

rvanheest replied:
You could, temporarily, but AFAIK accessibility and visibility are not the same!

If the keys (eg enum members) are the same, you could maybe rewrite this enumeration to an abstract class (sealed?) with `accessibility` and `visibility` texts and make instances (case classes or objects) for each accessibility type. See for a similar thing the [KeywordMapping in the License Creator](https://github.com/rvanheest-DANS-KNAW/easy-license-creator/blob/EASY-1039/src/main/scala/nl/knaw/dans/easy/license/KeywordMapping.scala) or [this assignment solution from the scala workshops](https://github.com/DANS-KNAW/course-scala/blob/master/src/main/scala/workshop1/assignments/UnitConversion.sc).

lindareijnhoudt replied:
i queried the fcrepo in april 2016. these were the values found there for the files:
accessible_to:
 {NONE, RESTRICTED_REQUEST, RESTRICTED_GROUP, KNOWN, ANONYMOUS}
visible_to:
 {ANONYMOUS, KNOWN, NONE}

i would suggest keeping them separate, so you can map them separately too.

jo-pol replied:
I'm afraid the keyword mapping and its alternative doesn't allow a unit test that guarantees an exhaustive key-set.

jo-pol committed:
refactored enum, applied it wherever it should

jo-pol commented on src/main/scala/nl/knaw/dans/easy/stage/fileitem/UserCategory.scala:
```diff
@@ -0,0 +1,69 @@
+  RESTRICTED_REQUEST, // a user that received permission to access the dataset
+  RESTRICTED_GROUP, // a user belonging to the same group as the dataset
+  NONE // none of the above
+  = Value
```
The legacy java code had different enums for VisibleTo and AccessibleTo, both "accidently" having the same names as instances.

I came to the conclusion accsibleTo and visibleTo are in fact file properties of the same class: a user category. "ANONYMOUS" and "KNOWN" properly reflect that semantic meaning. "RESTRICTED_GROUP" and "RESTRICTED_REQUEST" reflect different semantics, sadly it is too expensive (and certainly beyond this issue) to refactor them into something like "RECEIVED_PERMISSION" and "BELONGS_TO_GROUP" as we had to change data base values for a lot of files in the repository.

rvanheest replied:
Would it be useful to add some of this explanation to the code (as comment or scaladoc)?

jo-pol replied:
how?

PaulBoon replied:
UserCategory suggests a type of user while it is more like a user state. Maybe we should use AllowedUserState?

lindareijnhoudt replied:
i would like to rename (at some point in time) the AccessCategory to DatasetAccessCategory (or ~Type), because although it is in the dataset-package, this meaning is by no means obvious.
so i would suggest to rename this one to FileitemAccessCategory (or ~Type)

jo-pol commented:
Tested with 

```
./stageFileItem.sh -i easy-dataset:1 -p stageFileItem.sh -a RESTRICTED_REQUEST -l stageFileItem.sh -v RESTRICTED_GROUP --size 1 -d http://x target/testfilesdo
```

Don't understand why data stream location (-d) is mandatory while I see no trace of it in the created SDO

jo-pol commented:
Also tested with

```
./stageDataset.sh -d ddd -u uuu src/test/resources/dataset-bags/no-additional-license target/testDatasetSDO
```

neither DOI nor URN are configured as mandatory arguments but omitting them does cause runtime errors

jo-pol committed:
additional unit test

jo-pol commented:
In my opinion ready for a squashed merge.

jo-pol commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -69,10 +71,10 @@ object EasyStageDataset {
   }
 
-  private def createFileAndFolderSdos(dir: File, parentSDO: String)(implicit s: Settings): Try[Unit] = {
+  private def createFileAndFolderSdos(dir: File, parentSDO: String)(implicit s: Settings, rights: AccessCategory): Try[Unit] = {
```
@rvanheest 
I chose for implicit because it is an invariant for the recursion (unfold the hidden lines). Funny thing though, as a consequence I had to make setting more explicit at the initial call. Perhaps I should make it explicit for createFileSdo which doesn't pass it down any more.

rvanheest replied:
Personally I would prefer this access category to not be implicit. It would be clearer to have this as a 'normal' parameter, especially as the access category is not a singleton on project level (like Settings is).

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/fileitem/FileItemConf.scala:
```diff
@@ -53,14 +54,21 @@ class FileItemConf(args: Seq[String]) extends ScallopConf(args) {
+  def userCategory(default: UserCategory.Value) = singleArgConverter({s =>
+    if (s.trim.isEmpty) default else UserCategory.valueOf(s).get
+  })
+  def string(default: String) = singleArgConverter({s =>
```
1. Maybe you should give this function a better name. It is not really descriptive of what this function does.
2. In the License Creator project (as well as others) I have something similar like this inner lambda (see `package.scala`). Maybe it might be interesting to use something similar here as well. Then you get something like `s.toOption.getOrElse(default)`.
3. `userCategory(...)` and `string(...)` look fairly similar. Can you abstract over this?

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/fileitem/FileItemConf.scala:
```diff
@@ -115,8 +123,6 @@ class FileItemConf(args: Seq[String]) extends ScallopConf(args) {
   mutuallyExclusive(isMendeley, file)
 
-  validate(accessibleTo) (s => validateValue(s, accessCategories))
-  validate(visibleTo) (s => validateValue(s, accessCategories))
```
Why no more validation here for `accessibleTo` and `visibleTo`?

jo-pol replied:
validation is now implicit in the singleArgConverter userCategory

jo-pol commented:
Working on last comments...

jo-pol committed:
eliminated most implicit argument declarations

Could not get rid of it completely because
Util.LoadXML still needs settings.bagitDir, JSON/EMD need URN and DOI too

jo-pol committed:
less cryptic method name

jo-pol committed:
fedora out of settings allows more unit tests

debug logging now shows disciplines,
it used to log fedora url, user and ... password!

jo-pol committed:
created unit tests for StageDataset

jo-pol commented:
ready for merge again if reviewers agree

jo-pol committed:
dropped redundant fedora connection

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/EasyStageDataset.scala:
```diff
@@ -45,93 +47,100 @@ object EasyStageDataset {
+        mime <- readMimeType(getBagRelativePath(file).toString)
+        title <- readTitle(getBagRelativePath(file).toString)
+        _ <- EasyStageFileItem.createFileSdo(sdoDir, "objectSDO" -> parentSDO
+        )(FileItemSettings(
```
Misschien kun je dit beter leesbaar schrijven als op [line 125](https://github.com/DANS-KNAW/easy-stage-dataset/pull/76/files?diff=split#diff-0d13331279bc764a20980df7d8dcf552R125).

rvanheest commented on src/main/scala/nl/knaw/dans/easy/stage/fileitem/FileItemSettings.scala:
```diff
@@ -80,7 +76,31 @@ object FileItemSettings {
+    // no need to catch exceptions thrown by the constructor as the defaults take care of valid values
+    new FileItemSettings(
+      sdoSetDir = Some(sdoSetDir),
+      file = None,
```
Instead of all these `None`s, can't you give these as defaults to the case class itself? Worth looking at to make things less complicated.

jo-pol replied:
An attempt resulted in compiler error: `overloaded method value apply with alternatives:`

jo-pol committed:
last review comments
